modules/react-native-status/android/src/main/java/im/status/ethereum/module/JSCJail.java

```diff
@@ -0,0 +1,270 @@
+            public void web3sendAsync(final String payload, final JSValue callback) {
+                Thread thread = new Thread() {
+                    @Override
+                    public void run() {
+                        String result = Statusgo.CallRPC(payload);
+                        callback.toFunction().call(null, result);
+                    }
+                };
+
+                thread.start();
+            }
+        };
+        context.property("web3sendAsync", web3sendAsync);
+
+        JSFunction statusSendSignal = new JSFunction(context, "statusSendSignal") {
+            public void statusSendSignal(String data) {
+                JSONObject event = new JSONObject();
+                JSONObject signal = new JSONObject();
+                try {
+                    event.put("chat_id", chatId);
+                    event.put("data", data);
+                    signal.put("type", "jail.signal");
+                    signal.put("event", event);
+                } catch (JSONException e) {
+
+                }
+
+                module.signalEvent(signal.toString());
+            }
+        };
+        context.property("statusSendSignal", statusSendSignal);
+
+        JSFunction setTimeout = new JSFunction(context, "setTimeout") {
+            public String setTimeout(final JSValue callback, final int ms) {
+                return timer.setTimeout(callback, ms);
+            }
+        };
+        context.property("setTimeout", setTimeout);
+
+        JSFunction setInterval = new JSFunction(context, "setInterval") {
+            public String setInterval(final JSValue callback, final int ms) {
+                return timer.setInterval(callback, ms);
+            }
+        };
+        context.property("setInterval", setInterval);
+
+        JSFunction clearInterval = new JSFunction(context, "clearInterval") {
+            public void clearInterval(String id) {
+                timer.clearInterval(id);
+            }
+        };
+        context.property("clearInterval", clearInterval);
+
+        JSFunction statusLog = new JSFunction(context, "statusLog") {
+            public void statusLog(String data) {
+                Log.d("statusJSLog", data);
+            }
+        };
+        context.property("statusLog", statusLog);
+    }
+
+    private JSException jsexception;
+
+    private void checkException(String label) {
+        if (jsexception != null) {
+            jsexception = null;
+        }
+    }
+
+    @Override
+    public String parseJail(String chatId, String js) {
+        Cell cell = new Cell();
+        JSContext context = new JSContext();
+        cell.context = context;
+        cell.timer = new Timer();
+
+        context.setExceptionHandler(new JSContext.IJSExceptionHandler() {
+            @Override
+            public void handle(JSException exception) {
+                jsexception = exception;
+            }
+        });
+        String web3Js = "var statusSignals = {\n" +
```

Comment:
Can we put this in a resource file and load that in? Seems cleaner and less error prone to me

Reply:
And it's probably is safer but this code is kind of deeply coupled with injections which we are doing in both platforms and by putting it in completely different places we are making it a bit more effortful to find a clue when reading the code from both places (i mean native code and some hypothetical place where we put this js, like `resources` dir).

We had the same situation just yesterday when @jeluard tried to find where StatusBridge object comes from https://status-im.slack.com/archives/G6GCQH57X/p1510848455000385

So IMO both options are not perfect, in obj-c we at least are able to use macros for multiline strings, which makes this js a bit more readable, but java is java.